### PR TITLE
respect prometheus annotations

### LINKED ISSUE #31187: metrics merging: detect prometheus.io.scrape format
Due to escaping logic in prom, this is equivilent to prometheus.io/scrape (`__meta_kubernetes_pod_annotation_prometheus_io_path`). So we probably want to detect both. See https://github.com/openfaas/faas-netes/issues/714.

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

fixed  #31187

### CODE REVIEW

File: pkg/kube/inject/webhook.go [hunk 1]
```diff
@@ -608,10 +609,48 @@ func applyPrometheusMerge(pod *corev1.Pod, mesh *meshconfig.MeshConfig) error {
...
 
+var (
+	prometheusScrapeAnnotations = []string{
+		"prometheus.io/scrape",
+		"prometheus.io.scrape",
+	}
+
+	prometheusPortAnnotations = []string{
+		"prometheus.io/port",
+		"prometheus.io.port",
```
Comment (howardjohn) [inline pkg/kube/inject/webhook.go:626]: Prometheus has a variety of charecters that it escapes to the same thing. I don't know the fully list, but ideally we normalize the annotations in the same way they do.

For example `prometheus/io/port` would work with prometheus but this logic would not handle it

**Review by howardjohn — COMMENTED** [review]
Looks really close, thanks for the PR

File: pkg/kube/inject/webhook.go (hunk 1 above)
Comment (zirain) [reply pkg/kube/inject/webhook.go:626]: you're right, `prometheus/io/port` also will convert to `__meta_kubernetes_pod_annotation_prometheus_io_port`

**Comment by zirain** [comment]
according to  [SanitizeLabelName](https://github.com/prometheus/prometheus/blob/71a0f42331566a8849863d77078083edbb0b3bc4/util/strutil/strconv.go#L40)

@howardjohn please review again.

**Comment by zirain** [comment]
BTW, do we need a release notes?

File: pkg/kube/inject/webhook.go [hunk 2]
```diff
@@ -608,10 +610,45 @@ func applyPrometheusMerge(pod *corev1.Pod, mesh *meshconfig.MeshConfig) error {
...
+		return nil
 	}
+
 	return nil
 }
 
+var (
+	prometheusScrapeAnnotation = "prometheus_io_scrape"
+	prometheusPortAnnotation   = "prometheus_io_port"
+)
```
Comment (hzxuzhonghu) [inline pkg/kube/inject/webhook.go:622]: what's this?

Comment (zirain) [reply pkg/kube/inject/webhook.go:622]: prometheus will convert annotation to this format see [here](https://github.com/prometheus/prometheus/blob/71a0f42331566a8849863d77078083edbb0b3bc4/discovery/kubernetes/pod.go#L205)

`prometheus.io/scrape` `prometheus.io.scrape` `prometheus-io/scrape` have the same meaning in Prometheus

Comment (howardjohn) [reply pkg/kube/inject/webhook.go:622]: Putting that comment in the code would be ideal for future reference

File: pkg/kube/inject/webhook.go [hunk 3]
```diff
@@ -608,10 +610,45 @@ func applyPrometheusMerge(pod *corev1.Pod, mesh *meshconfig.MeshConfig) error {
 		pod.Annotations["prometheus.io/port"] = targetPort
 		pod.Annotations["prometheus.io/path"] = "/stats/prometheus"
 		pod.Annotations["prometheus.io/scrape"] = "true"
```
Comment (howardjohn) [inline pkg/kube/inject/webhook.go:612]: What if a user sets `prometheus/io/path: foo`. Then we add `prometheus.io/path: /stats/prometheus`. Then won't we end up with *both*, at which point prometheus may pick a random one?

We may need to clear out all variants and then set ours?

**Review by howardjohn — COMMENTED** [review]
Nice!

File: pkg/kube/inject/webhook.go (hunk 3 above)
Comment (zirain) [reply pkg/kube/inject/webhook.go:612]:
```golang
        for k, v := range pod.Annotations {
		ln := strutil.SanitizeLabelName(k)
		ls[model.LabelName(podAnnotationPrefix+ln)] = lv(v)
		ls[model.LabelName(podAnnotationPresentPrefix+ln)] = presentValue
	}
```
I think prometheus will take a random one due to range from a map 

I prefer to merge istio metric to the path which's already settled in annotations, wdyk @howardjohn

Comment (zirain) [reply pkg/kube/inject/webhook.go:612]:
```
We may need to clear out all variants and then set ours?
```
seems reasonable, Is there something told end users about this auto-merge behavior?

Comment (howardjohn) [reply pkg/kube/inject/webhook.go:618]: We do say https://istio.io/latest/docs/ops/integrations/prometheus/#option-1-metrics-merging they will be overriden so it seems ok to override? WDYT

Comment (zirain) [reply pkg/kube/inject/webhook.go:621]: updated

File: pkg/kube/inject/webhook.go [hunk 4]
```diff
@@ -605,13 +612,84 @@ func applyPrometheusMerge(pod *corev1.Pod, mesh *meshconfig.MeshConfig) error {
...
+		if scrape, err := strconv.ParseBool(val); err == nil {
+			return scrape
+		}
+	}
+
+	return true
+}
+
+func clearPrometheusAnnotations(pod *corev1.Pod) {
+	prometheusAnnotations := sets.NewString(
```
Comment (hzxuzhonghu) [inline pkg/kube/inject/webhook.go:645]: This can be moved out

**Review by hzxuzhonghu — COMMENTED** [review]
LGTM

File: releasenotes/notes/36959.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+kind: bug-fix
+area: installation
+
+# issue is a list of GitHub issues resolved in this note.
+issue:
+  - 31187
+
+releaseNotes:
+- |
+  **Fixed** detect `prometheus.io.scrape` format when merging metrics
```
Comment (ericvn) [inline releasenotes/notes/36959.yaml:11]: Maybe:
```suggestion
  **Fixed** detection of `prometheus.io.scrape` annotations when merging metrics.
```
```suggestion
  **Fixed** detect `prometheus.io.scrape` format when merging metrics
```

Comment (ericvn) [inline releasenotes/notes/36959.yaml:11]: Maybe.
```suggestion
  **Fixed** detection of `prometheus.io.scrape` annotations when merging metrics.
```

**Review by ericvn — APPROVED** [review]

**Comment by zirain** [comment]
ping @howardjohn @hzxuzhonghu

File: pkg/kube/inject/webhook.go [hunk 6]
```diff
@@ -36,6 +37,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/serializer"
 	kjson "k8s.io/apimachinery/pkg/runtime/serializer/json"
 	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/kube-openapi/pkg/util/sets"
```
Comment (howardjohn) [inline pkg/kube/inject/webhook.go:40]: nit: use `pilot/pkg/util/sets`

**Review by howardjohn — APPROVED** [review]
:+1:

**Pull request merged**